src/impy/common.py

```diff
@@ -756,32 +771,16 @@ def _define_default_fs_particles(self):
         info(5, "Setting following particles to be stable:", impy_config["stable_list"])
 
         for pdgid in impy_config["stable_list"]:
-            self.set_stable(pdgid)
+            self._set_stable(pdgid, True)
 
         if impy_config["pi0_stable"]:
-            self.set_stable(111)
+            self._set_stable(111, True)
 
-    def __call__(self, nevents):
-        """Generator function (in python sence)
-        which launches the underlying event generator
-        and returns its the result (event) as MCEvent object
-        """
-        retry_on_rejection = impy_config["retry_on_rejection"]
-        # Initialize counters to prevent infinite loops in rejections
-        ntrials = 0
-        nremaining = nevents
-        while nremaining > 0:
-            self._update_event_kinematics()
-            if self.generate_event() == 0:
-                self.nevents += 1
-                yield self._event_class(self)
-                nremaining -= 1
-                ntrials += 1
-            elif retry_on_rejection:
-                info(10, "Rejection occured. Retrying..")
-                ntrials += 1
-                continue
-            elif ntrials > 2 * nevents:
-                raise Exception("Things run bad. Check your input.")
-            else:
-                info(0, "Rejection occured")
+        self._set_final_state_particles_called = True
+
+    @contextmanager
```

Comment:
That's cool. So what it does it to avoid altering kinematics during cross section readout. Very nice.

Reply:
Yeah, since we use this in several places, it is useful to have a context manager for this that changes the variable and puts the original back in place, to avoid code repetition and possible bugs.

Reply:
Context managers are easy to write with `@contextmanger` and `yield`, so once you learn this you find many places to use it :).